fix: refact `rln-relay` and post sync test

darshankabariya opened the pull request:
# Description
This refactor was necessary because we’ve already transitioned to a non-synchronous approach, and all sync-related code is being deprecated.

Over time, GroupManager became bloated with tightly coupled and complex logic. This PR initiates a refactor to improve the clarity, maintainability, and future extensibility of the RLN relay protocol.

Changes

- [x] Restructured the GroupManager hierarchy and introduced offchainGroupManager to replace staticGroupManager.
- [x] Deprecated the static manager class in favor of the offchainManager interface. (As suggested by @jm-clius, both on-chain and off-chain trees may exist in the future; currently, OffchainManager is just an interface.)
- [x] Removed all zerokit ffi for the merkle tree along with its associated proof generation and verification wrappers.
- [x] Deleted all unit tests related to the old merkle tree implementation.
- [x]  Update all effected with onchain anvail

## Issue

- https://github.com/waku-org/nwaku/issues/3483

darshankabariya commented:
Hi @Ivansete-status, @jm-clius, @stubbsta —
The PR is now ready for review.
A few test cases are taking longer than expected — I’m still looking into that.
However, all the core logic changes for improving the RLN relay protocol are complete.

jm-clius reviewed (commented):
Thanks! I've just eyeballed the changes for now - I think this is a good step towards simplification. So far one-ish question below.

darshankabariya commented on waku/waku_rln_relay/group_manager/off_chain/group_manager.nim:
Yes, you got it right. Actually, I couldn’t find any use for the StaticGroupManager class, so I deprecated it along with its unit tests. As you noticed, I introduced the OffchainGroupManager interface—this idea came from your suggestion that we might support both offchain and onchain features in the future. I’ve left it unimplemented for now.

darshankabariya commented on waku/waku_rln_relay/group_manager/on_chain/group_manager.nim:
```diff
@@ -202,20 +197,21 @@ proc trackRootChanges*(g: OnchainGroupManager) {.async: (raises: [CatchableError
       if rootUpdated:
         if g.membershipIndex.isNone():
-          error "membershipIndex is not set; skipping proof update"
+          debug "membershipIndex is not set; skipping proof update"
```
It’s no longer required because earlier registering RLN membership was mandatory, but now users can run the code without membership. so removed it.

darshankabariya commented on tests/waku_rln_relay/test_rln_serde.nim:
```diff
@@ -33,7 +34,7 @@ suite "RLN Relay v2: serde":
     check expectedLeaf == leafRes.value.inHex()
 
-  test "proofGen: generates a valid zk proof":
+  xasyncTest "proofGen: generates a valid zk proof":
```
[tests/waku_rln_relay/test_rln_serde.nim] is no longer needed. It was testing proof generation in isolation, but after moving to the on-chain tree, proof generation requires running a node and registering membership. Since it’s no longer possible to generate proofs fully off-chain, this test case is obsolete. deleting whole file.

darshankabariya commented on tests/waku_rln_relay/test_rln_serde.nim:
```diff
@@ -22,7 +23,7 @@ func defaultRateCommitment*(): RateCommitment =
 
 suite "RLN Relay v2: serde":
-  test "toLeaf: converts a rateCommitment to a valid leaf":
+  xasyncTest "toLeaf: converts a rateCommitment to a valid leaf":
```
same !!

darshankabariya commented on tests/waku_rln_relay/rln/waku_rln_relay_utils.nim:
```diff
@@ -1,23 +1,54 @@
+    epochSizeSec: uint64 = 1,
+    treePath: string = genTempPath("rln_tree", "waku_rln_relay"),
+    index: MembershipIndex = MembershipIndex(0),
+): WakuRlnConfig =
+  let wakuRlnConfig = WakuRlnConfig(
```
It’s already in the test file, so why do we need to mention it explicitly ?

darshankabariya commented on tests/node/test_wakunode_legacy_lightpush.nim:
```diff
@@ -158,7 +158,7 @@ suite "RLN Proofs as a Lightpush Service":
 
   suite "Lightpush attaching RLN proofs":
-    asyncTest "Message is published when RLN enabled":
+    xasyncTest "Message is published when RLN enabled":
```
Unable testcase !!

darshankabariya commented on tests/waku_rln_relay/test_rln_group_manager_offchain.nim:
```diff
@@ -0,0 +1,44 @@
+    credentials.add(generateCredentials(rlnInstance))
+  return credentials
+
+suite "Offchain group manager":
```
Yes, I agree — we no longer need the off-chain testcase since there’s nothing left to test. but also there no harm at all, that why not delete.

fryorcraken commented on tests/waku_rln_relay/rln/waku_rln_relay_utils.nim:
```diff
@@ -1,23 +1,54 @@
-  let proof = rlnPeer.groupManager.generateProof(input, epoch, 0).valueOr:
+  # Fetch Merkle proof either when a new root was detected *or* when the cache is empty.
+  if rootUpdated or manager.merkleProofCache.len == 0:
+    let proofResult = waitFor manager.fetchMerkleProofElements()
```
a test is a top logic layer as it uses the library as a binary.

darshankabariya replied:
Got it !!

fryorcraken commented on waku/waku_rln_relay/group_manager/group_manager_base.nim:
```diff
@@ -119,58 +119,23 @@ method stop*(g: GroupManager): Future[void] {.base, async.} =
-
-  var rootBuffer: Deque[MerkleNode]
-  let overflowedRoots = slideRootQueue(g.validRoots, rootAfterUpdate)
-  if overflowedRoots.len > 0:
-    for root in overflowedRoots:
-      discard rootBuffer.slideRootQueue(root)
-  rootBuffer
+  return g.indexOfRoot(root) >= 0
```
Such a one liner method sounds like unnecessary indirection. You are not "validating" anything, so it actually create confusion.

darshankabariya replied:
In this snippet, I removed slideRootQueue and simplified validateRoot. I’m not sure what specific confusion you’re referring to.

fryorcraken commented on waku/waku_rln_relay/constants.nim:
```diff
@@ -37,18 +37,18 @@ const
 const
-  StaticGroupSize* = 10000
-  #  StaticGroupKeys is a static list of 10000 membership keys in the form of (identity trapdoor, identity nullifier, identity secret hash, identity commitment)
+  OffchainGroupSize* = 10000
+  #  OffchainGroupKeys is a static list of 10000 membership keys in the form of (identity trapdoor, identity nullifier, identity secret hash, identity commitment)
   #  keys are created locally, using createMembershipList proc from waku_rln_relay_utils module, and the results are hardcoded in here
   #  this list is temporary and is created to test the performance of waku-rln-relay for the static groups
   #  in the later versions, this static hardcoded group will be replaced with a dynamic one
```
Can this be moved under the `tests` folder?

Or, if it is to be used by a binary, just move it near said binary.

darshankabariya replied:
No, it’s not used in tests — it’s only part of the binary.

fryorcraken commented on waku/waku_rln_relay/rln_relay.nim:
```diff
@@ -49,45 +49,11 @@ type RlnRelayConf* = object of RootObj
   treePath*: string
   epochSizeSec*: uint64
   userMessageLimit*: uint64
+  ethPrivateKey*: Option[string]
```
Is there a type we can use to prevent accidental printing of the `string`?

darshankabariya replied:
Thanks for pointing that out. I think we should add some logic to the log interface, specifically for ethPrivateKey.

fryorcraken commented on waku/waku_rln_relay/constants.nim:
```diff
@@ -37,18 +37,18 @@ const
 const
-  StaticGroupSize* = 10000
-  #  StaticGroupKeys is a static list of 10000 membership keys in the form of (identity trapdoor, identity nullifier, identity secret hash, identity commitment)
+  OffchainGroupSize* = 10000
```
`Static` was a better name. But this is test data, so I would name it like that.

darshankabariya commented on waku/waku_rln_relay/group_manager/off_chain/group_manager.nim:
```diff
@@ -0,0 +1,54 @@
+  groupKeys*: seq[IdentityCredential]
+  groupSize*: uint
+
+template initializedGuard*(g: OffchainGroupManager): untyped =
+  discard
```
Thanks for pointing out, I’ve updated all functions to return an unimplemented error.

darshankabariya commented on tests/waku_rln_relay/utils_onchain.nim:
```diff
@@ -570,17 +570,16 @@ proc setupOnchainGroupManager*(
   let testTokenAddress = (await deployTestToken(privateKey, acc, web3)).valueOr:
-    assert false, "Failed to deploy test token contract: " & $error
-    return
+    raise
+      newException(CatchableError, "Failed to deploy test token contract: " & $error)
```
I updated it with the idea of unification. I noticed we mostly use exceptions in our code, so I replaced asserts with exceptions. But now I realize this is for binary tests, so exceptions may not be a good idea.

darshankabariya commented:
> LGTM! I'd make sure we remove all the tests that are not needed and keep the needed ones. F.e., it seems we should not skip the following in `tests/waku_rln_relay/rln/test_wrappers.nim
> And also, CI green :D

Hi @Ivansete-status , thank you for the approval. I’ve addressed all your suggestions and updated the changes accordingly. All CI checks are green now and tests have been updated as well. If you see this before merging the PR, kindly have a look.

jm-clius reviewed (changes requested):
Thanks! I'm ready to approve once I understand why the `OffchainGroupManager` exists at all. :) Will be happy to have a call if I'm missing some context or my comments are unclear.

darshankabariya commented on waku/waku_rln_relay/group_manager/off_chain/group_manager.nim:
```diff
@@ -0,0 +1,64 @@
+
+export group_manager_base
+
+# Placeholder for future OffchainGroupManager re-implementation
+type OffchainGroupManager* = ref object of GroupManager
```
thanks for review. main objective for offchainGM is to serve as a future perspective, so that when we revisit the offchainTree, we can leverage the currently unimplemented offchainGM functionality. But I understand your point — we can remove offchainGM for now as well.

darshankabariya commented on waku/waku_rln_relay/rln_relay.nim:
```diff
@@ -447,11 +413,11 @@ proc mount(
 
   if not conf.dynamic:
```
Yes!! we also need to remove that along with offchainGM.

jm-clius commented on waku/waku_rln_relay/rln_relay.nim:
```diff
@@ -445,38 +411,25 @@ proc mount(
   let rlnInstance = createRLNInstance(tree_path = conf.treePath).valueOr:
     return err("could not create RLN instance: " & $error)
 
-  if not conf.dynamic:
```
I think this config item must be marked in the external config as deprecated (and then removed within two releases). Presumably it will not break any existing deployments and the default is `true` in any case?

darshankabariya replied:
It's really good idea.

jm-clius reviewed (approved):
Thanks! Two comments below, but no need to re-request review once addressed.
